opt: Support building a memo with aggregates and GROUP BY

rytaft opened the pull request:
This is the second commit of several needed to build a "memo"
structure from a parsed SQL query. The memo is the primary data
structure in the query optimizer, used for maintaining the forest of
query plan trees that the optimizer considers as part of its search
for the optimal plan.

This commit includes support for building a memo for queries
which include aggregates or a `GROUP BY` clause. The scoping rules
for grouping are complex, so the complexity in this commit is mostly
for the purpose of enforcing these rules. The test cases have been
adapted from the aggregate test in logictests.

Future commits will add support for `DISTINCT` and `HAVING`, as well
as `VALUES`, `UNION` and complex joins.

This code is based on code originally written by @petermattis and
@andy-kimball in the opttoy project.

Release note: None

rytaft committed:
opt: Support building a memo with aggregates and GROUP BY

This is the second commit of several needed to build a "memo"
structure from a parsed SQL query. The memo is the primary data
structure in the query optimizer, used for maintaining the forest of
query plan trees that the optimizer considers as part of its search
for the optimal plan.

This commit includes support for building a memo for queries
which include aggregates or a GROUP BY clause. The scoping rules
for grouping are complex, so the complexity in this commit is mostly
for the purpose of enforcing these rules. The test cases have been
adapted from the aggregate test in logictests.

Future commits will add support for DISTINCT and HAVING, as well
as VALUES, UNION and complex joins.

This code is based on code originally written by @petermattis and
@andy-kimball in the opttoy project.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/22297)
<!-- Reviewable:end -->

rytaft commented:
Sorry for the change, but I found a bug in my first version.  I've fixed the bug and pushed the latest changes.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-:-L4INfYpr14pLu5G_TXj:b-nw21rz)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Just skimming these changes to stay informed. Nit: is every test data file in `build/testdata/` going to be prefixed by `build-`?

---

Review status: 0 of 5 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/build/scope.go, line 93 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4IOGTC7Af4Pcal3jSs:-L4IOGTDwY-rDtwKpP83:b-pvr39h) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b741752f693a9458479541bb3ab3d5dacd6084/pkg/sql/opt/build/scope.go#L93)):*
> ```Go
> 	expr, _ = tree.WalkExpr(s, expr)
> 	// Important - we call the expr.TypeCheck method rather than the TypeCheck
> 	// function to avoid folding of constant literals.
> ```

Why don't you want to fold constant literals?

---

*[pkg/sql/opt/build/testdata/build-aggregate, line 17 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4IOZkcrpA-ge60VyLf:-L4IOZkcrpA-ge60VyLg:bqzb3as) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b741752f693a9458479541bb3ab3d5dacd6084/pkg/sql/opt/build/testdata/build-aggregate#L17)):*
> ```
> 
> build
> SELECT MIN(1), MAX(1), COUNT(1), SUM_INT(1), AVG(1), SUM(1), STDDEV(1), VARIANCE(1), BOOL_AND(true), BOOL_AND(false), XOR_AGG(b'\x01') FROM t.kv
> ```

I believe the query can be multiple lines to avoid really long lines like this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-:-L4ISWeB5VxyF3sGgD_u:b-n12659)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks for the review! I also was starting to think I should get rid of the `build-` prefix -- done.

---

Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/build/scope.go, line 93 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4IOGTC7Af4Pcal3jSs:-L4IUu0e6h0vHWzweWvZ:b-83g6zt) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b741752f693a9458479541bb3ab3d5dacd6084/pkg/sql/opt/build/scope.go#L93)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Why don't you want to fold constant literals?
</blockquote></details>

It changes the semantics for `GROUP BY`, apparently. 

For example, `SELECT COUNT(*) FROM t.kv GROUP BY 1+2` is different from `SELECT COUNT(*) FROM t.kv GROUP BY 3`. The first query is grouping on the constant expression "1+2", while the second one is grouping on the third column in the select list (it throws an error since there is only one column).

---

*[pkg/sql/opt/build/testdata/aggregate, line 17 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4IOZkcrpA-ge60VyLf:-L4IUr_m3o6klY8mWmXC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b741752f693a9458479541bb3ab3d5dacd6084/pkg/sql/opt/build/testdata/aggregate#L17)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I believe the query can be multiple lines to avoid really long lines like this.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-:-L4IVXsmgqikT1NKMdq9:b-1db543)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/scope.go, line 93 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4IOGTC7Af4Pcal3jSs:-L4JC2tJUh1uG--UIhcS:b-cwht4) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b741752f693a9458479541bb3ab3d5dacd6084/pkg/sql/opt/build/scope.go#L93)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

It changes the semantics for `GROUP BY`, apparently. 

For example, `SELECT COUNT(*) FROM t.kv GROUP BY 1+2` is different from `SELECT COUNT(*) FROM t.kv GROUP BY 3`. The first query is grouping on the constant expression "1+2", while the second one is grouping on the third column in the select list (it throws an error since there is only one column).
</blockquote></details>

Does current Cockroach get this wrong? I worry that not doing the constant folding will change our typing behavior as constant folding is an integral part of our type inference. Let's expand this comment to capture your reasoning here and add a TODO as I'm pretty sure we'll need to revisit this. Cc @nvanbenschoten and @knz. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

knz commented:
This distinction occurs before constant folding. It's pretty well specified in the pg docs and the SQL specs.

Same for ORDER BY.

BTW i believe this was initially implemented wrong in crdb and I seem to recall I fixed it.

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

knz commented:
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/scope.go, line 93 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22297#-L4IOGTC7Af4Pcal3jSs:-L4JRmiVBUiNIcl8oow8:b8luyat) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b741752f693a9458479541bb3ab3d5dacd6084/pkg/sql/opt/build/scope.go#L93)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Does current Cockroach get this wrong? I worry that not doing the constant folding will change our typing behavior as constant folding is an integral part of our type inference. Let's expand this comment to capture your reasoning here and add a TODO as I'm pretty sure we'll need to revisit this. Cc @nvanbenschoten and @knz. 
</blockquote></details>

ow

the handling here is incorrect I believe.

The test for e.g. GROUP BY 3 must match exactly a literal numeric constant. It's a *lexical* check which must occur before any handling by TypeCheck. In particular `GROUP BY (3)` is not to be handled like `GROUP BY 3` - only the latter of the two gets special treatment.

Once the code has determined that the lexical structure of the parameter is not a literal numeric constant, `TypeCheck` can be called as for any other expression.

Note that ORDER BY has the same rule, and *also* another even weirder rule that takes place *before* this one. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/scope.go, line 93 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4IOGTC7Af4Pcal3jSs:-L4JjHsf6--slvBSUn-A:b-8k64d3) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b741752f693a9458479541bb3ab3d5dacd6084/pkg/sql/opt/build/scope.go#L93)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ow

the handling here is incorrect I believe.

The test for e.g. GROUP BY 3 must match exactly a literal numeric constant. It's a *lexical* check which must occur before any handling by TypeCheck. In particular `GROUP BY (3)` is not to be handled like `GROUP BY 3` - only the latter of the two gets special treatment.

Once the code has determined that the lexical structure of the parameter is not a literal numeric constant, `TypeCheck` can be called as for any other expression.

Note that ORDER BY has the same rule, and *also* another even weirder rule that takes place *before* this one. 
</blockquote></details>

I tried this on Postgres 9.6, and GROUP BY (3) is handled the same way as GROUP BY 3. In both cases, it groups by the 3rd column in the result. It doesn't appear to be a pure lexical check in Postgres:

http://sqlfiddle.com/#!17/315e3/15

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/scope.go, line 93 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4IOGTC7Af4Pcal3jSs:-L4JqdnqMMr-0LYyKEE5:bbtixgw) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b741752f693a9458479541bb3ab3d5dacd6084/pkg/sql/opt/build/scope.go#L93)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes, you're right, as a special case the parentheses are unwrapped too. I had forgotten that.
</blockquote></details>

It's still not good to do type check because e.g. `group by (3::int)` (which would simplify during type checking) is not the same thing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/opt/build/builder.go, line 302 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22297#-L4MFlfL8ctZHPe-D7Ne:-L4MFlfL8ctZHPe-D7Nf:b-cz8ppm) ([raw file](https://github.com/cockroachdb/cockroach/blob/faa60ff473857e3f9ce8e9a19a96569fcb582e3a/pkg/sql/opt/build/builder.go#L302)):*
> ```Go
> // this scalar expression.
> //
> // If hasAgg is true, the second return value, varsUsed, contains any variables
> ```

Why is varUsed not a set? Do we want the same ColumnIndex potentially appearing multiple times? If yes, the comment should explain that.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 0 of 7 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/build/builder.go, line 302 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4MFlfL8ctZHPe-D7Ne:-L4b-TVF0SgudFikICdh:b-kn2p4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/faa60ff473857e3f9ce8e9a19a96569fcb582e3a/pkg/sql/opt/build/builder.go#L302)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Why is varUsed not a set? Do we want the same ColumnIndex potentially appearing multiple times? If yes, the comment should explain that.
</blockquote></details>

I added `varsUsed` to `scope` per @andy-kimball's suggestion and added a comment there to explain why it needs to be a slice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks for the reviews.  I believe I've fixed the issue with the lexical check on GROUP BY, and I reorganized the code a bit to keep `varsUsed` inside `scope`. Please take another look - thanks!

---

Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/build/scope.go, line 93 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4IOGTC7Af4Pcal3jSs:-L4bLLsO09VGgwl5VfUL:b-p0tukb) ([raw file](https://github.com/cockroachdb/cockroach/blob/21b741752f693a9458479541bb3ab3d5dacd6084/pkg/sql/opt/build/scope.go#L93)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

It's still not good to do type check because e.g. `group by (3::int)` (which would simplify during type checking) is not the same thing.
</blockquote></details>

Thanks for the info! I think I've fixed this now - please take another look.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-:-L4bN2hFRxgu93GAyOZz:bbyzmw4)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: 0 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 671 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4bVsxjMkywKIGZgcjs:-L4bVsxjMkywKIGZgcjt:b-yug6g4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/builder.go#L671)):*
> ```Go
> 	if col != -1 {
> 		groupBy = selects[col].Expr
> 		label = string(selects[col].As)
> ```

yep, looks good

---

*[pkg/sql/opt/build/builder.go, line 709 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4bW5IPIJYxiNn7RFp5:-L4bW5IPIJYxiNn7RFp6:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/builder.go#L709)):*
> ```Go
> // Here "1" refers to the first item in the SELECT list, "a". The returned index
> // is 0.
> func colIndex(numOriginalCols int, expr tree.Expr, context string) int {
> ```

:+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 0 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 671 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4bVsxjMkywKIGZgcjs:-L4bbGU314Yg62JCOUWn:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/builder.go#L671)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

yep, looks good
</blockquote></details>

Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Great stuff! LGTM, just some nits

---

Review status: 0 of 10 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/build/builder.go, line 302 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c2DKsfimO5rixtSVA:-L4c2DKsfimO5rixtSVB:bvxl9f6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L302)):*
> ```Go
> 		}
> 		out = b.factory.ConstructVariable(b.factory.InternPrivate(t.index))
> 		return
> ```

We try to not use naked returns, `return out` or just `return b.factory..`

---

*[pkg/sql/opt/build/builder.go, line 386 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c2KIgtuKJlBwNNkji:-L4c2KIhq3Z5LrI1L7Mj:b-1gj3kt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L386)):*
> ```Go
> 	}
> 
> 	return
> ```

return out

---

*[pkg/sql/opt/build/builder.go, line 446 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c2Z3IwTX-nxtdX7cL:-L4c2Z3IwTX-nxtdX7cM:b-1gj3kt) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L446)):*
> ```Go
> 	}
> 
> 	return
> ```

return out

---

*[pkg/sql/opt/build/builder.go, line 531 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c5hhp5iQCzMlan1Rf:-L4c5hhp5iQCzMlan1Rg:b-4ltidz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L531)):*
> ```Go
> 		// Any aggregate columns that were discovered would have been appended
> 		// to the end of the grouping scope.
> 		aggCols := groupingsScope.cols[len(groupingsScope.cols)-len(groupingsScope.groupby.aggs):]
> ```

we do this in a couple of places, maybe we want an `aggCols` method

---

*[pkg/sql/opt/build/builder.go, line 629 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c3OCPHmGTDD5IZQcW:-L4c3OCPHmGTDD5IZQcX:b-34h72h) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L629)):*
> ```Go
> ) (groupings []opt.GroupID, outScope *scope) {
> 	if groupBy == nil {
> 		return
> ```

return nil, nil

---

*[pkg/sql/opt/build/builder.go, line 641 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c3PUIq1Z23TxPqQ4x:-L4c3PUIq1Z23TxPqQ4y:btlvyrj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L641)):*
> ```Go
> 	}
> 
> 	return
> ```

return groupings, outScope

---

*[pkg/sql/opt/build/builder.go, line 930 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c6TcfLT3gIp5bOfO_:-L4c6TcgPIhNHVWTLoNn:bah15s7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L930)):*
> ```Go
> }
> 
> // buildVariableProjection builds a set of memo groups that represent the given
> ```

Isn't there just one memo group always?

---

*[pkg/sql/opt/build/scope.go, line 70 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c3lyy64-W5ajHl9Bb:-L4c3lyz7JgmtPc-Wey5:b-ht8lb) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/scope.go#L70)):*
> ```Go
> 	inGroupingContext bool
> 
> 	// varsUsed is only utilized when inGroupingContext is true.
> ```

Nice comment

---

*[pkg/sql/opt/build/scope.go, line 225 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c7hfzRb7_zR15Zdf2:-L4c7hfzRb7_zR15Zdf3:b-jsbfw4) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/scope.go#L225)):*
> ```Go
> 
> 	s.groupby.inAgg = false
> 	return
> ```

return refScope

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-:-L4c6CuDJBUl3mJKXTh2:b-26ksiz)*
<!-- Sent from Reviewable.io -->

rytaft commented:
TFTR! Updated.

---

Review status: 0 of 10 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/sql/opt/build/builder.go, line 302 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c2DKsfimO5rixtSVA:-L4cGRipjgTXvT5E3Yn6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L302)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We try to not use naked returns, `return out` or just `return b.factory..`
</blockquote></details>

Done.

---

*[pkg/sql/opt/build/builder.go, line 386 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c2KIgtuKJlBwNNkji:-L4cGSMQwQyc3gfrJiJv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L386)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

return out
</blockquote></details>

Done.

---

*[pkg/sql/opt/build/builder.go, line 446 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c2Z3IwTX-nxtdX7cL:-L4cG_6-FOCjjRBjR6Uh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L446)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

return out
</blockquote></details>

Done.

---

*[pkg/sql/opt/build/builder.go, line 531 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c5hhp5iQCzMlan1Rf:-L4cGcFbNF6APmgpRi8w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L531)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

we do this in a couple of places, maybe we want an `aggCols` method
</blockquote></details>

Done.

---

*[pkg/sql/opt/build/builder.go, line 629 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c3OCPHmGTDD5IZQcW:-L4cGf1wPCm0WBX8RfRn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L629)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

return nil, nil
</blockquote></details>

Done.

---

*[pkg/sql/opt/build/builder.go, line 641 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c3PUIq1Z23TxPqQ4x:-L4cGh3J173QwnDleCIM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L641)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

return groupings, outScope
</blockquote></details>

Done.

---

*[pkg/sql/opt/build/builder.go, line 930 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c6TcfLT3gIp5bOfO_:-L4cGixt2ZH-hr2-3pVT:b-qu6n94) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/builder.go#L930)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Isn't there just one memo group always?
</blockquote></details>

Good point. Done.

---

*[pkg/sql/opt/build/scope.go, line 70 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c3lyy64-W5ajHl9Bb:-L4cGnnM2owgwPr7KhTZ:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/scope.go#L70)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Nice comment
</blockquote></details>

Thanks!

---

*[pkg/sql/opt/build/scope.go, line 225 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4c7hfzRb7_zR15Zdf2:-L4cGr888W5lxWgTCXbp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e525bcaa662ac990ee50c0b27993a795ba53780c/pkg/sql/opt/build/scope.go#L225)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

return refScope
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-:-L4cGu0us7HpClDGoDFB:b-p0fj95)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 10 files reviewed at latest revision, 17 unresolved discussions, some commit checks broke.

---

*[pkg/sql/opt/build/builder.go, line 632 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4clJTJvu0ZwbNUO7Vf:-L4clJTJvu0ZwbNUO7Vg:bi7qkep) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/builder.go#L632)):*
> ```Go
> 	}
> 
> 	// After GROUP BY, variables from inScope are hidden
> ```

But what about the subquery case?
```
SELECT a.x
FROM a
WHERE EXISTS(SELECT a.x FROM b GROUP BY b.x)
```
Unfortunately subqueries make everything more complex. I think your changes fixed some bugs in opttoy, but they introduced new bugs in subquery handling. The `scope.push()` calls are necessary to keep ancestor scopes in subquery cases.

---

*[pkg/sql/opt/build/builder.go, line 759 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4b_Hx5jugjM13osydP:-L4b_Hx5jugjM13osydQ:bkgvbv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/builder.go#L759)):*
> ```Go
> 	exprTransformCtx := transform.ExprTransformContext{}
> 	for _, sel := range selects {
> 		if exprTransformCtx.AggregateInExpr(sel.Expr, b.semaCtx.SearchPath) {
> ```

Does this helper handle subqueries properly?

---

*[pkg/sql/opt/build/scope.go, line 68 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4b_WROJsJcDbk7JeDL:-L4b_WROJsJcDbk7JeDM:b-m413v1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/scope.go#L68)):*
> ```Go
> 	// true, varsUsed will be utilized to enforce scoping rules. See the comment
> 	// below for more details.
> 	inGroupingContext bool
> ```

Is inGroupingContext true iff groupingsScope is non-nil? If so, can't inGroupingContext be a helper function that returns `groupingsScope != nil`? If not, then maybe fix up the comment to make it more clear what you mean by "active", because my intuition would be that "active" means it's not nil.

---

*[pkg/sql/opt/build/scope.go, line 173 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4cR8VvyFGiB9wb2ZyR:-L4cR8VvyFGiB9wb2ZyS:b-6f6zxx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/scope.go#L173)):*
> ```Go
> func (s *scope) findGrouping(grouping opt.GroupID) *columnProps {
> 	for i, g := range s.groupby.groupings {
> 		if g == grouping {
> ```

Based on our discussion with Raphael, I don't think this will work, since groups are normalized. We'll need to determine equivalence using AST nodes rather than groups. That's just as well, since I think it's best to not rely upon the properties of the output nodes in order to check semantics of the input nodes.

---

*[pkg/sql/opt/build/scope.go, line 195 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4cglGenLwKhU4L6Cno:-L4cglGenLwKhU4L6Cnp:bfevsnf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/scope.go#L195)):*
> ```Go
> 
> 		if curr.groupby.groupingsScope != nil {
> 			// The aggregate will be added to the innermost groupings scope.
> ```

What if you have a query like this:
```
SELECT t1.a
FROM t1
GROUP BY t1.a
HAVING EXISTS
(
  SELECT MAX(t1.a) FROM t2 GROUP BY t2.b
)
```
The original opttoy code set the refScope when compiling column references, since that's when you find out which groupings scope that aggregate is referencing. The new code will just always force aggregates to refer to the innermost scope. Is that a deliberate limitation for now since we don't yet support subqueries?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 0 of 10 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[pkg/sql/opt/build/builder.go, line 632 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4clJTJvu0ZwbNUO7Vf:-L4fnCzrsJG-dIcphNSc:bgfuz1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/builder.go#L632)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

But what about the subquery case?
```
SELECT a.x
FROM a
WHERE EXISTS(SELECT a.x FROM b GROUP BY b.x)
```
Unfortunately subqueries make everything more complex. I think your changes fixed some bugs in opttoy, but they introduced new bugs in subquery handling. The `scope.push()` calls are necessary to keep ancestor scopes in subquery cases.
</blockquote></details>

Good point.  Per our offline discussion, I've changed it to inScope.replace() so that the parent scope is hidden, but ancestor scopes are still available.

---

*[pkg/sql/opt/build/builder.go, line 759 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4b_Hx5jugjM13osydP:-L4fkokkswhyYWoE3GPa:b-n5llhu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/builder.go#L759)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Does this helper handle subqueries properly?
</blockquote></details>

I believe so. It uses the Visitor interface to walk the expression tree, and it explicitly does not recurse into subqueries.

---

*[pkg/sql/opt/build/scope.go, line 68 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4b_WROJsJcDbk7JeDL:-L4fl6bAxeGy9o2dKSn3:b-m0wmjm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/scope.go#L68)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Is inGroupingContext true iff groupingsScope is non-nil? If so, can't inGroupingContext be a helper function that returns `groupingsScope != nil`? If not, then maybe fix up the comment to make it more clear what you mean by "active", because my intuition would be that "active" means it's not nil.
</blockquote></details>

I've removed it and created a helper function.  I had to move some things around to make sure that `groupingsScope` is only created in a grouping context.

---

*[pkg/sql/opt/build/scope.go, line 173 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4cR8VvyFGiB9wb2ZyR:-L4flaMdKU92EVgE7WAD:b-xeop0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/scope.go#L173)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Based on our discussion with Raphael, I don't think this will work, since groups are normalized. We'll need to determine equivalence using AST nodes rather than groups. That's just as well, since I think it's best to not rely upon the properties of the output nodes in order to check semantics of the input nodes.
</blockquote></details>

You are right - this query works in both CockroachDB and PG:
`select a+b from t group by a+b`
but this query doesn't:
`select a+b from t group by b+a`

(I think supporting the second query would be a feature rather than a bug, but I suppose we should try to stick as closely as possible to the current functionality)

I'll work on doing the AST comparison...

---

*[pkg/sql/opt/build/scope.go, line 195 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4cglGenLwKhU4L6Cno:-L4fmMbyy-w86gI5TanP:blncaqg) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/scope.go#L195)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

What if you have a query like this:
```
SELECT t1.a
FROM t1
GROUP BY t1.a
HAVING EXISTS
(
  SELECT MAX(t1.a) FROM t2 GROUP BY t2.b
)
```
The original opttoy code set the refScope when compiling column references, since that's when you find out which groupings scope that aggregate is referencing. The new code will just always force aggregates to refer to the innermost scope. Is that a deliberate limitation for now since we don't yet support subqueries?
</blockquote></details>

I think I was a bit confused by the original opttoy code since it seems the `refScope` functionality wasn't fully hooked up.  So I didn't deliberately limit the `refScope`, but now I see that this will need to change to support subqueries. Perhaps that support can wait until the next PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
*[pkg/sql/opt/build/scope.go, line 173 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4cR8VvyFGiB9wb2ZyR:-L4fpOxndjf43ITviDYr:bklzs0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/scope.go#L173)):*
> I think supporting the second query would be a feature rather than a bug

I would think so too.. I don't think we should expend effort into making such queries error out.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

rytaft commented:
I've addressed Andy's comments and done a bit of refactoring to remove redundant code. Please take another look. Thanks!

---

Review status: 0 of 11 files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[pkg/sql/opt/build/builder.go, line 759 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4b_Hx5jugjM13osydP:-L4h9LDjZWsOlr3jJpPc:b-kv3fyn) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/builder.go#L759)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I believe so. It uses the Visitor interface to walk the expression tree, and it explicitly does not recurse into subqueries.
</blockquote></details>

As discussed at the meeting today, we *do* need to recurse into subqueries since aggregates in correlated subqueries should be included. I've added a TODO to address this in a future PR.

---

*[pkg/sql/opt/build/scope.go, line 173 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4cR8VvyFGiB9wb2ZyR:-L4hBnjafAUPUCxc-ojR:bif57tc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/scope.go#L173)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

> I think supporting the second query would be a feature rather than a bug

I would think so too.. I don't think we should expend effort into making such queries error out.
</blockquote></details>

This function is no longer being used for query validation, only for enabling reuse of column expressions.

---

*[pkg/sql/opt/build/scope.go, line 195 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4cglGenLwKhU4L6Cno:-L4hC_eqmJAebVG0aW1U:bn5gurl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff9cadb0478cab058bf86ce944f304ebf7ea4dd2/pkg/sql/opt/build/scope.go#L195)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

I think I was a bit confused by the original opttoy code since it seems the `refScope` functionality wasn't fully hooked up.  So I didn't deliberately limit the `refScope`, but now I see that this will need to change to support subqueries. Perhaps that support can wait until the next PR?
</blockquote></details>

Added a TODO to fix this in a future PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-:-L4hDSjli1ow3C5eylR1:b-3xn8qh)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
Review status: 0 of 11 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[pkg/sql/opt/build/builder.go, line 382 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4iRNfBZv1iSzyV3ZW3:-L4iRNfCda_4u_pif2hX:b-w7i3gs) ([raw file](https://github.com/cockroachdb/cockroach/blob/d697de3ef3b0ac74959c2f0c639fe9ab869134a3/pkg/sql/opt/build/builder.go#L382)):*
> ```Go
> 	// a GROUP BY expression, truncate varsUsed.
> 	if inGroupingContext {
> 		if _, ok := inScope.groupby.groupStrs[symbolicExprStr(scalar)]; ok {
> ```

Won't this generate a string for every expression tree within a grouping context? So if we have several large expression trees in a select list containing 100 expression ops within them, we'd be creating 100 strings. And each string would in effect regenerate all the strings in its subtree, making the overall operation exponential in terms of number of times the leaves would be regenerated by the formatter.

Is this the way that Cockroach is currently doing it as well?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

knz commented:
Op 06-02-18 om 23:23 schreef Andy Kimball:

> Is this the way that Cockroach is currently doing it as well?

Yes although I doubt this was ever an issue in practice given the
limited support for subqueries in scalar contexts.

-- 
Raphael 'kena' Poss

rytaft commented:
Review status: 0 of 11 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/sql/opt/build/builder.go, line 382 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4iRNfBZv1iSzyV3ZW3:-L4kaIzAVLu1ZU9DgopQ:b-wn4n98) ([raw file](https://github.com/cockroachdb/cockroach/blob/d697de3ef3b0ac74959c2f0c639fe9ab869134a3/pkg/sql/opt/build/builder.go#L382)):*
> Yes although I doubt this was ever an issue in practice given the limited support for subqueries in scalar contexts.

I've added a TODO to come back and fix this, in the interest of not delaying this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297)*
<!-- Sent from Reviewable.io -->

andy-kimball commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 11 files reviewed at latest revision, 14 unresolved discussions, some commit checks broke.

---

*[pkg/sql/opt/build/builder.go, line 382 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-L4iRNfBZv1iSzyV3ZW3:-L4klfG-1jaI7rPSspQJ:b-ufdqt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d697de3ef3b0ac74959c2f0c639fe9ab869134a3/pkg/sql/opt/build/builder.go#L382)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

> Yes although I doubt this was ever an issue in practice given the limited support for subqueries in scalar contexts.

I've added a TODO to come back and fix this, in the interest of not delaying this PR.
</blockquote></details>

OK, seems like if it hasn't caused problems so far, we're OK to proceed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-:-L4klndGymYS1t4bg1fD:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks for the review!

---

Review status: 0 of 11 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/22297#-:-L4kmWppN2K4WxEcjiLu:b-9qdoz3)*
<!-- Sent from Reviewable.io -->
